Adding reporting ID capability in reactor module

yjung-anl opened the pull request:
closes #19217

yjung-anl committed:
adding reporting ID for cartesian lattice, refs idaholab#19217

yjung-anl committed:
adding reporting ID for hexagonal lattice, refs idaholab#19217

yjung-anl committed:
adding reporting ID for plane, refs idaholab#19217

yjung-anl committed:
adding moosedoc for reporting ID functions, refs idaholab#19217

cticenhour reviewed (dismissed):
There are quite a few excess commits here that have already been merged into MOOSE. I would rebase your reactor module commits on top of the current devel branch to get rid of these.

GiudGiud commented on modules/reactor/src/meshgenerators/PlaneIDGenerator.C:
```diff
@@ -0,0 +1,115 @@
+defineLegacyParams(PlaneIDGenerator);
+
+InputParameters
+PlaneIDGenerator::validParams()
+{
+  InputParameters params = MeshGenerator::validParams();
+
+  params.addRequiredParam<MeshGeneratorName>("input", "The mesh we want to modify");
+  params.addRequiredParam<std::vector<Real>>("z_layers", "Z coordinates of axial layers");
+  params.addParam<std::vector<unsigned int>>("z_sublayers",
+                                             "The number of layers for each z-layers");
+  params.addRequiredParam<std::string>("id_name", "Name of Integer ID set");
+  params.addClassDescription(
+      "This PlaneIDGenerator source code is to assigns plane reporting ID for existing 3D meshes");
+
+  return params;
```
this is very specific (Z planes, only planes supported etc), lets have something more general

yjung-anl replied:
`PlaneIDGenerator` was updated to take `x`, `y` or `z` as a plane axis option as:

- Example 1
```
  [pin3d_id]
    type = PlaneIDGenerator
    input = 'pin3d'
    layer_coordinates = '0.0 5.0 10.0 15.0'
    num_sublayers = ' 1 2 1'
    layer_axis = 'z'
    id_name = 'plane_id'
  []
```

- Example 2
```
  [grid]
    type = CartesianMeshGenerator
    dim = 2
    dx = '5.0'
    ix = '2'
    dy = '5.0 10.0 '
    iy = '1 4'
  []

  [plane_id_gen]
    type = PlaneIDGenerator
    input = 'grid'
    layer_coordinates = '0.0 5.0 15.0'
    num_sublayers = ' 1 2'
    layer_axis = 'y'
    id_name = 'plane_id'
  []
```

GiudGiud commented on modules/reactor/src/meshgenerators/PlaneIDGenerator.C:
```diff
@@ -0,0 +1,115 @@
+    std::vector<Real> base_z_layers = getParam<std::vector<Real>>("z_layers");
+    std::vector<unsigned int> z_sublayers = getParam<std::vector<unsigned int>>("z_sublayers");
+    if (base_z_layers.size() != z_sublayers.size() + 1)
+      mooseError("Sizes of 'z_layers' and 'z_sublayers' disagree");
```
```suggestion
      paramError("z_layers, "Sizes of 'z_layers' and 'z_sublayers' disagree");
```

GiudGiud commented on modules/reactor/src/meshgenerators/PlaneIDGenerator.C:
```diff
@@ -0,0 +1,115 @@
+      for (unsigned int j = 0; j < z_sublayers[i]; ++j)
+        _layers.push_back(_layers.back() + layer_size);
+    }
+    for (auto z : _layers)
+    {
+      _console << z << std::endl;
+    }
```
```suggestion
```

GiudGiud commented on modules/reactor/src/vectorpostprocessors/IntegralReportingIDVPP.C:
```diff
@@ -0,0 +1,180 @@
+      ids = getAllElemIDs(id_index);
+    }
+    _reporting_id_index.push_back(id_index);
+    std::map<dof_id_type, dof_id_type> id_map;
+    std::map<dof_id_type, dof_id_type> id_inverse_map;
```
I think the data types can be improved here

also I d be surprised if none of these two maps don't exist already. Check libmesh doxygen just in case

yjung-anl replied:
Updated this VPP not to use those two data types.

GiudGiud commented on modules/reactor/src/vectorpostprocessors/IntegralReportingIDVPP.C:
```diff
@@ -0,0 +1,180 @@
+    _var_integrals.push_back(&p);
+  }
+
+  dof_id_type i = 0;
+  for (auto id_type : _reporting_id)
+  {
+    auto & p = declareVector("Level-" + std::to_string(i) + "-" + id_type);
+    p.resize(unique_ids.size());
+    _var_reporting_ids.push_back(&p);
+    ++i;
+  }
+  dof_id_type ipos = 0;
+  for (dof_id_type uid : unique_ids)
+  {
+    dof_id_type uid0 = uid;
+    for (dof_id_type i = _n_reporting_id - 1; i > 0; --i)
+    {
+      dof_id_type n = _reporting_id_map[i].size();
+      auto & id_inverse_map = _reporting_id_inverse_map[i];
+      auto it = id_inverse_map.find(uid0 % n);
+      dof_id_type id = it->second;
+      (*_var_reporting_ids[i])[ipos] = id;
+      uid0 = uid0 / n;
+    }
+    auto & id_inverse_map = _reporting_id_inverse_map[0];
+    auto it = id_inverse_map.find(uid0);
+    dof_id_type id = it->second;
+    (*_var_reporting_ids[0])[ipos] = id;
+    _reporting_id_to_pos[uid] = ipos;
+    ++ipos;
+  }
+}
```
add comments

GiudGiud commented on modules/reactor/test/tests/meshgenerators/reporting_id/cartesian_id/assembly_reporting_id_cell.i:
```diff
@@ -0,0 +1,62 @@
+[Mesh]
+  [pin1]
+    type = ConcentricCircleMeshGenerator
+    num_sectors = 4
```
can you drop to less sectors or less rings? gold file size are proportional to that

GiudGiud commented on modules/reactor/test/tests/meshgenerators/reporting_id/cartesian_id/tests:
```diff
@@ -0,0 +1,39 @@
+    type = 'Exodiff'
+    input = 'assembly_reporting_id_cell.i'
+    exodiff = 'assembly_reporting_id_cell_out.e'
+    requirement ='The system shall generate pin reporting IDs for cartesian assembly with assign_type=cell (default)'
```
don't mention parameters in requirement, explain what it is doing instead.
```suggestion
    requirement = 'The system shall generate pin reporting IDs for cartesian assembly with assign_type=cell (default)'
```
fix spacing too (see others as well)

GiudGiud commented on modules/reactor/test/tests/meshgenerators/reporting_id/cartesian_id/tests:
```diff
@@ -0,0 +1,39 @@
+    type = 'Exodiff'
+    input = 'assembly_reporting_id_pattern.i'
+    exodiff = 'assembly_reporting_id_pattern_out.e'
+    requirement ='The system shall generate pin reporting IDs for cartesian assembly with assign_type=pattern'
```
same

GiudGiud commented on modules/reactor/test/tests/meshgenerators/reporting_id/cartesian_id/tests:
```diff
@@ -0,0 +1,39 @@
+    type = 'Exodiff'
+    input = 'assembly_reporting_id_manual.i'
+    exodiff = 'assembly_reporting_id_manual_out.e'
+    requirement ='The system shall generate pin reporting IDs for cartesian assembly with assign_type=manual'
```
same

GiudGiud commented on modules/reactor/test/tests/meshgenerators/reporting_id/depletion_id/tests:
```diff
@@ -0,0 +1,18 @@
+    type = 'Exodiff'
+    input = 'depletion_id_exclude_material.i'
+    exodiff = 'depletion_id_exclude_material_out.e'
+    requirement ='The system shall generate plane depletion IDs for 2d core with exclude_material_id option'
```
```suggestion
    requirement ='The system shall generate plane depletion IDs for 2d core with the possibility to skip some materials'
```

GiudGiud commented on modules/reactor/test/tests/vectorpostprocessors/tests:
```diff
@@ -0,0 +1,11 @@
+    type = 'CSVDiff'
+    input = 'reporting_id_vpp.i'
+    csvdiff = 'reporting_id_vpp_out_integral_0001.csv'
+    requirement ='The system shall generate vector processed data based on reporting ID for 2d problem'
```
```suggestion
    requirement ='The system shall be able to integrate variables based on the extra element ID'
```

GiudGiud commented on modules/reactor/test/tests/meshgenerators/reporting_id/hexagonal_id/assembly_reporting_id_pattern.i:
```diff
@@ -0,0 +1,65 @@
+    duct_sizes = '2.4 2.5'
+    duct_intervals = '1 1'
+    id_name = 'pin_id'
+    assign_type = 'pattern'
```
these 3 input files (maybe only 2) can be merged.
then in the tests spec, pass a command line argument:
cli_args = 'Mesh/assembly/assign_type=pattern'

GiudGiud reviewed (changes requested):
Move the image file to large_media instead of checking it in the MOOSE repository

The depletionIDGenerator could just be a ParsedIDGenerator, where extra element IDs are parsed to generate a new element ID. Then we can have a general framework object instead of something specific to these two mesh generators in the module

The PlaneIDGenerator could be with planes in any direction. It should actually also be a ParsedIDGenerator, taking a general expression, or some sort of FunctionIDGenerator, taking a function, to generate IDs based on the spatial location. 

@YaqiWang I think you should take a look at this PR to see if there s some synergy that can be created for your work on adding CMFD indices as extra element ID. 

Tests can be simplified using cli_args. And it would be great to keep the exodus file sizes down as much as possible.

eshemon commented:
> Move the image file to large_media instead of checking it in the MOOSE repository
> 
> The depletionIDGenerator could just be a ParsedIDGenerator, where extra element IDs are parsed to generate a new element ID. Then we can have a general framework object instead of something specific to these two mesh generators in the module
> 
> The PlaneIDGenerator could be with planes in any direction. It should actually also be a ParsedIDGenerator, taking a general expression, or some sort of FunctionIDGenerator, taking a function, to generate IDs based on the spatial location.
> 
> @YaqiWang I think you should take a look at this PR to see if there s some synergy that can be created for your work on adding CMFD indices as extra element ID.
> 
> Tests can be simplified using cli_args. And it would be great to keep the exodus file sizes down as much as possible.

Hi @GiudGiud - thanks for the review. @yjung-anl is modifying some of your suggestions now. I would like to request that we keep the DepletionIDGenerator and PlaneIDGenerator as is for now. I do hear your point that these could be generalized and placed further down in the framework, but by generalizing them and not using the terminology/orientation that is typical for reactor analysts, we may lose some "useability" for the end user. One alternative would be creation of the more general routine in the framework (directly in meshgenerators, not in modules), and then a "wrapper" type of routine that sits up in the Reactor module -  but that's a little out of scope for our current program funding which is focus on enabling reactor analysis with NEAMS applications. Would it be acceptable to keep everything as is for now, or should we have a chat/more discussion? I want to strike the right balance between generality and useability for our specific stakeholders.

friedmud commented:
@eshemon is spot on about generality vs. usefulness - there will always be a trade off there.

Personally: I don't like the idea of _everything_ being a `Parsed*`.  Taken to the extreme there would be no need to write code - we would just write entire applications in the input file.

I'm not suggesting that @GiudGiud is completely misguided here... just that there is a cutoff where overgeneralization is not helpful.

The way I understand what's going to happen going forward here is that these are foundational objects that other reactor mesh generators will be built on top of and they will be utilized as sub-generators in a more compact syntax.  I think it makes sense for these to exist as _code_.

I think that @eshemon and her team are doing a great job of identifying redundancy where they can and utilizing inheritance and sub-generators to keep down code duplication.  I would suggest to them that they continue to look for opportunities to do that - but that these objects should be allowed as they are.

GiudGiud commented:
You're contributing 5 objects here, 3 of which could be generalized to be helpful to others. I am mostly trying to raise awareness at this point, this PR isn't held up by this right now. 

I don't think the sub-generator argument is valid here. Nothing stops you from having a parsed expression as an argument to a sub-generator in your code. If anything, this is the best reason to have advanced objects since the user won't even see them!

I think I'm ok with DepletionIDGenerator stay that way, and I ll create an issue to get a ParsedExtraIDGenerator separately in the framework. There is quite a lot of case-specific behavior with the assign_type that would be a little annoying to add in a parsed expression

The ReportingIDVPP is a general object already. It has the wrong name, but that s about all you need to change to make it a framework object. Something like ExtraIDVariableIntegrationVPP (or VectorPostprocessor but lots of letters at this point) and a `git mv` and you're good to go!

The PlaneIDGenerator is also a framework object at heart, even as it is with no flexibility. I would love a IsoSurfaceExtraIDGenerator that takes in a parsed (quadratic most often) equation. If you don't want to do that, at least add a normal parameter to a PlaneExtraIDGenerator. There are horizontal reactor designs, at least 2 advanced ones I know of.

YaqiWang commented:
I will take a look.

eshemon commented:
@friedmud and @GiudGiud Thanks for the additional discussion. @GiudGiud - you make some good points. I appreciate the specific suggestions you brought up.

@yjung-anl and I briefly discussed and we agree that the ReportingIDVPP should be down in the framework. He's going to remove it from this PR and make a separate one. 

As for PlaneIDGenerator, we can extend to x-y-z direction vectors which are relevant for the reactor applications you mentioned (but not a general vector). This would be a low maintenance update for us. If it's restricted to these vectors, should it stay in Reactor module? Or move down to the framework?

One thing that's a little confusing (from a NEAMS application perspective) is how to point new users to a collective set of tools that they can sift through/read documentation on. I was hoping to keep a lot of stuff in the same place  (e.g. in modules, or in the framework) just to aid the users, but I understand that keeping things down in the framework is more desirable in general for MOOSE. Appreciate any suggestions/guidance on where to put things.

YaqiWang commented:
This link shows an example as @loganharbour suggested: https://griffin-dev.hpc.inl.gov/site/radiation_transport-user-manual/mesh.html#mesh-system

permcody commented on modules/reactor/src/vectorpostprocessors/IntegralReportingIDVPP.C:
```diff
@@ -0,0 +1,180 @@
+/****************************************************************/
+/*               DO NOT MODIFY THIS HEADER                      */
+/*  MAMMOTH: - Radiation Kinetics and Reactor Physics Software  */
+/*                                                              */
+/*           (c) 2016 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
```
@YaqiWang - Who is the original author on this code? We probably should get permission from them, and give them proper attribution, etc.

yjung-anl replied:
Sorry, it is my mistake.  I will change it properly

yjung-anl commented on modules/reactor/doc/content/source/meshgenerators/CartesianIDPatternedMeshGenerator.md:
```diff
@@ -0,0 +1,44 @@
+
+- `cell` (default):  Assign unique  IDs for each component in the lattice in sequential order.
+
+- `pattern`:  Assign IDs for each input component type.
```
I see your point. Perhaps, do you have any suggestion on a keyword to replace `pattern`?

yjung-anl commented on modules/reactor/src/meshgenerators/DepletionIDGenerator.C:
```diff
@@ -0,0 +1,123 @@
+      std::pair<dof_id_type, dof_id_type> ids(id1, id2);
+      elem->set_extra_integer(depletion_id, map_depletion_id[ids]);
+    }
+    base_id = depletion_id;
```
In this algorithm, in the first stage, determine a depletion ID set by searching unique combinations of the material ID and the first reporting ID  defined in `id_name`. From the second step, a new combinations is searched by combining the depletion ID from the previous step and the next reporting ID defined in `id_name`. 
Initially, `base_id` name is set `material_id`. Once the first iteration is done, it should be replaced to `depletion_id`. I think this algorithm is confusing.

yjung-anl replied:
If just copy the material ID to the depletion ID, prior to enter this loop. Then, `base_id = depletion_id` is not needed and easy to understand.

GiudGiud commented:
I m ok with the IntegralReportingIDVPP being renamed, placed in framework and part of this PR. I think it s a great capability.

For the planeID, maybe we can repurpose a Griffin mesh generator, the one to generate coarse grids, instead. The planeIDGenerator is currently just a special case of it with nx=1 and ny=1.

yjung-anl commented on modules/reactor/src/meshgenerators/DepletionIDGenerator.C:
```diff
@@ -0,0 +1,123 @@
+        auto it = merge_base_id.find(ids.second);
+        if (it != merge_base_id.end())
+        {
+          new_id = it->second;
```
I understand. If I update the algorithm given in the below comment, this part can be simpler.

YaqiWang reviewed (changes requested):
I looked into `CartesianIDPatternedMeshGenerator` and `DepletionIDGenerator` more carefully in this review and just browsed through the others. Overall this is the right direction of assigning IDs and a great new capability. Depending on how we think element IDs can be more useful, we can make `DepletionIDGenerator` more general. If I make the code, especially in MOOSE, I will make it a general one. Just in documentation, I give a concrete example on how it can be used to assign depletion IDs. The PlaneIDGenerator is too specific. If we are not careful, we will be facing an issue of lots of mesh generators. I'd rather users to remember the control parameters in a general generator than the names of specific generators.

YaqiWang commented on modules/reactor/src/meshgenerators/CartesianIDPatternedMeshGenerator.C:
```diff
@@ -0,0 +1,176 @@
+      }
+    }
+  }
+  return integer_ids;
```
We should have an inline comment about this somewhere.

yjung-anl commented on modules/reactor/src/meshgenerators/DepletionIDGenerator.C:
```diff
@@ -0,0 +1,123 @@
+      if (it == unique_ids.end())
+        unique_ids.insert(std::pair<unsigned int, unsigned int>(id1, id2));
+    }
+    comm().set_union(unique_ids);
```
Here, check whether a combination of `(id1, id2)` is defined in `unique_ids`. If not defined, added to `unique_id`. So I think there is no chance of error here.

yjung-anl commented on modules/reactor/src/meshgenerators/DepletionIDGenerator.C:
```diff
@@ -0,0 +1,123 @@
+DepletionIDGenerator::validParams()
+{
+  InputParameters params = MeshGenerator::validParams();
+  params.addRequiredParam<MeshGeneratorName>(
```
>>Suggest to make this optional, when missing, this generator will be assigning depletion IDs based on material IDs with exclusion of certain materials. Sounds to me a simple extension.

You mean:

```
params.addRequiredParam<std::vector<std::string>>("id_names", "Extra integer id names");
```
this one as optional?

yjung-anl replied:
it can be easily done

YaqiWang replied:
Yes.

yjung-anl replied:
I updated it. I will push the updated depletion ID generator, first. I think it has a better shape now. 
Thank you for your comments @YaqiWang @GiudGiud.

friedmud commented:
@eshemon and I just talked about this and most of the suggestions by @GiudGiud will be followed :+1:

For now - the PlaneIDGenerator should stay a "plane"... but with a `direction` of `x y z` just like the `LayeredBase` does.  DepletionID will be generalized (but may have a specialization to make it simpler to use for reactors too).  ReportingIDVPP will be slightly generalized - with the added idea that we need to make _all_ UserObjects restrictable to element IDs at some point in the future.

Thank you very much for all of the ideas everyone - all good brainstorming.  I'm sure, that in time we can do all the things that are here - but there is also some practicality to consider about moving on to the next task (as it is, this is wrapping up working from last year).

friedmud commented:
One more thing: If you feel really strongly that you have a good idea for extended / more general capability... please make an Issue for it and reference it here so we don't lose track of it!

eshemon commented:
![Capture](https://user-images.githubusercontent.com/83663101/139990280-b51c6c2f-b580-4835-a3ab-0e8f98df01fd.PNG)
Couldn't copy/paste a table, so trying a screenshot. This is the path forward as I understand it (same thing Derek said).

GiudGiud commented:
I'd suggest:
- ExtraIDVariableIntegralVPP. Some people like to integrate material properties or functions
- Plane**Extra**IDGenerator. But only if you prefer to do this over bringing over the Griffin object Yaqi and I mentioned. A CartesianGridExtraIDGenerator (or whatever the name was, but I like that name) definitely provides much more flexibility. And the coding work is already done there!
- the plan for depletionIDGenerator sounds good to me. 

I like the idea of adding a new interface like "ExrtaIDRestrictable" a lot. At the end of the day, a block ID is also an extra ID. This is clearly more work that needs not be in this PR.

friedmud commented:
@GiudGiud You're exactly right `ExtraIDRestrictable` is what I was thinking.

Question: do we need to be clear that these are _element_ IDs?  @roystgnr , do we have extra IDs on nodes too?

loganharbour commented:
Both exist. It’s just extra storage on the DofObject. You define them similarly on the mesh.

eshemon commented:
1) So should we change ExtraID -> ExtraElementID everywhere in the naming?
2) We will take @GiudGiud naming suggestion for the VPP.
3) @yjung-anl looked into source code for Griffin's CartesianGridID and it is not currently flexible enough for our needs. It appears to require a 3D cartesian mesh input and also puts extra (x,y) pin/assembly IDs on the mesh. This is not our mesh of interest, and we also don't want these extra ids - they are assigned elsewhere, purposefully. The existing PlaneIDGenerator is agnostic of the mesh in the other two dimensions which is needed for every reactor type except cartesian LWR. If someone wants to create a more general version of CartesianGridID, move it to the framework, and allow only Z-dimension to be input, we will later update our PlaneIDGenerator to use that. Given this, would you like us to put PlaneIDGenerator in Reactor module, or framework?

roystgnr commented:
> @roystgnr , do we have extra IDs on nodes too?

Any DofObject can be assigned extra data.  With the most natural API for doing that assignment, the MeshBase interfaces, you give the mesh names (one at a time or one vector of names at a time) and corresponding data types (typically dof_id_type, though we'll do scary casting if you ask for something else, and we have test coverage for Real) you want preallocated on new elements, and/or you give it independent names and corresponding data types you want preallocated on new nodes, and the mesh will do so and will also hang on to the concatenated sets to let you do name->index lookups later.  In theory you might even be able to pick out individual nodes and elements afterward to add even more data to on top of what the mesh added, but that's not really a supported capability right now.

YaqiWang commented:
> * ExtraIDVariableIntegralVPP. Some people like to integrate material properties or functions

Do we need to carry `VPP` in the name? I suggest to remove it. Also I know now we have `MeshGenerator` in names of several mesh generators, which is required when we have meshes and modifiers but no longer needed.

YaqiWang commented:
`CartesianGridID` in Griffin is different. It could be deprecated by possibly using the mesh generators in reactor module. What I mean `CartesianIDMeshGenerator` is to simply extend `PlainIDGenerator`.

YaqiWang commented:
> @GiudGiud You're exactly right `ExtraIDRestrictable` is what I was thinking.
> 

Just want to have a comment here that using element IDs makes us able to keep the number of mesh blocks low, otherwise MOOSE can take significant memory for doing block restriction. When we do element ID restriction, we need to be mindful to keep this in mind.

GiudGiud commented:
ExtraElementID is so wordy.
Anyone has a better name?

friedmud commented:
I think we can just go with `ExtraID`... it will actually be applicable to both things.

I think `PlaneExtraIDMeshGenerator` is good (and all that's needed - it doesn't need to be extended at this point)

`VectorPostprocessor` should be at the end of the other name.

The reason for this is collisions.  `VariableIntegral` could be a `Postprocessor`, `VectorPostprocessor`, or `AuxKernel` (or maybe something else).  `VariableIntegralVectorPostprocessor` is unique.

yjung-anl commented on modules/reactor/doc/content/source/meshgenerators/CartesianIDPatternedMeshGenerator.md:
```diff
@@ -0,0 +1,44 @@
+- `manual`: Assign IDs based on user-defined mapping defined in [!param](/Mesh/CartesianIDPatternedMeshGenerator/id_pattern).
+
+The default numbering scheme starts at 0 in the upper left hand corner of the grid and increments by 1 as the grid is traversed left to right, top to bottom.
+The name of the reporting ID is provided through id_name depending on the hierarchical level of component.
```
id_name is replaced to:
```
[!param](/Mesh/CartesianIDPatternedMeshGenerator/id_name)
```

yjung-anl commented on modules/reactor/src/meshgenerators/CartesianIDPatternedMeshGenerator.C:
```diff
@@ -0,0 +1,176 @@
+
+std::vector<dof_id_type>
+CartesianIDPatternedMeshGenerator::getPatternIntegerIDs() const
+{
```
`exclude_id` is designed to only work for `assign_type = cell`.
Error-out was added if not 'assign_type = cell`

yjung-anl commented on modules/reactor/src/meshgenerators/CartesianIDPatternedMeshGenerator.C:
```diff
@@ -0,0 +1,176 @@
+
+std::vector<dof_id_type>
+CartesianIDPatternedMeshGenerator::getManualIntegerIDs() const
+{
```
exclude_id is designed to only work for assign_type = cell.
Error-out was added if not 'assign_type = cell`

miaoyinb commented on modules/reactor/src/meshgenerators/SimpleHexagonGenerator.C:
```diff
@@ -125,5 +125,7 @@ SimpleHexagonGenerator::generate()
         _external_boundary_name;
   }
 
+  mesh->prepare_for_use();
```
`ReplicatedMesh::stitching_helper()` checks `libmesh_assert_equal_to (_is_prepared, other_mesh._is_prepared);` and throws an error in DBG mode if this mesh is "not prepared". This can be either put here or before stitching in `PatternedHexMG`.

yjung-anl commented:
This is still working version.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/876553/) on f1e6baa wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19218/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19218/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 30685d5e180a2a32649318687ea16dc3dc2b3c97`

yjung-anl commented on modules/reactor/doc/content/source/meshgenerators/HexIDPatternedMeshGenerator.md:
```diff
@@ -0,0 +1,47 @@
+Similarly, the assembly reporting IDs (`assembly_id`) are assigned in the core construction process.
+
+Certain regions can be excluded from being labeled with an ID, for example dummy regions that will later be deleted.
+This can be accommodated by listing mesh objects in the [!param](/Mesh/HexIDPatternedMeshGenerator/exclude_id) input parameter.
```
```java
  [core]
    type = HexIDPatternedMeshGenerator
    inputs = 'assembly dummy'
    pattern_boundary = none
    pattern = '  1 0 1;
                0 0 0 0;
               1 0 0 0 1;
                0 0 0 0;
                 1 0 1'
    assign_type = 'cell'
    id_name = 'assembly_id'
    exclude_id = 'dummy'
  []
```

In this example, input `dummy` (id = 1 in pattern input) will be excluded from the reporting id numbering.

yjung-anl commented:
In this push, `PlaneIDGenerator` was updated to take a plane direction as an input parameter.

loganharbour commented on modules/reactor/include/vectorpostprocessors/IntegralReportingIDVPP.h:
```diff
@@ -0,0 +1,47 @@
+#pragma once
+
+#include "ElementVariableVectorPostprocessor.h"
+
+#include <ostream>
```
```suggestion
```

loganharbour commented on modules/reactor/src/meshgenerators/HexIDPatternedMeshGenerator.C:
```diff
@@ -0,0 +1,217 @@
+
+registerMooseObject("ReactorApp", HexIDPatternedMeshGenerator);
+
+defineLegacyParams(HexIDPatternedMeshGenerator);
+
```
```suggestion
```

loganharbour commented on modules/reactor/src/meshgenerators/HexIDPatternedMeshGenerator.C:
```diff
@@ -0,0 +1,217 @@
+defineLegacyParams(HexIDPatternedMeshGenerator);
+
+InputParameters
+HexIDPatternedMeshGenerator::validParams()
```
There is a significant amount of duplication between this and the Cartesian equivalent. We need to get around this

yjung-anl replied:
@loganharbour 
Perhaps, I can create a new class `IDPatternedMeshGeneratorBase` which contains the duplicated functions.

loganharbour commented on modules/reactor/src/meshgenerators/PlaneIDGenerator.C:
```diff
@@ -0,0 +1,109 @@
+{
+  std::unique_ptr<MeshBase> mesh = std::move(_input);
+  if (mesh->mesh_dimension() < _axis_index + 1)
+    mooseError(
+        "MeshGenerator ", name(), " in PlaneIDGenerator must operate on a proper layer axis");
```
Should be a `paramError`

loganharbour commented on modules/reactor/src/vectorpostprocessors/IntegralReportingIDVPP.C:
```diff
@@ -0,0 +1,165 @@
+  // list of extra_integer id
+  std::set<dof_id_type> ids;
+
+  std::set<ExtraElementIDName> _reporting_id_copy(_reporting_id.begin(), _reporting_id.end());
```
```suggestion
  std::set<ExtraElementIDName> reporting_id_set(_reporting_id.begin(), _reporting_id.end());
```

`_` prefix is reserved for member vars

loganharbour commented on modules/reactor/src/vectorpostprocessors/IntegralReportingIDVPP.C:
```diff
@@ -0,0 +1,165 @@
+void
+IntegralReportingIDVPP::initialize()
+{
+  for (unsigned int ivar = 0; ivar < _nvar; ++ivar)
+    for (size_t i = 0; i < (*_var_integrals[ivar]).size(); ++i)
+      (*_var_integrals[ivar])[i] = 0;
```
```suggestion
  for (auto & var_integral : _var_integrals)
    std::fill(var_integral->begin(), var_integral->end(), 0);
```

loganharbour commented on modules/reactor/src/vectorpostprocessors/IntegralReportingIDVPP.C:
```diff
@@ -0,0 +1,165 @@
+registerMooseObject("ReactorApp", IntegralReportingIDVPP);
+
+InputParameters
+IntegralReportingIDVPP::validParams()
```
Can this object not be more general? I think something collecing values based on element integer may be valid in the farmework

loganharbour replied:
Also curious if you guys have ever looked at the Reporter system. It enables outputting arbitrary JSON. May prove useful once more output is desired.

yjung-anl replied:
I am not aware of the `reporter system`. I will take a look. Thanks.

loganharbour commented on modules/reactor/src/vectorpostprocessors/IntegralReportingIDVPP.C:
```diff
@@ -0,0 +1,165 @@
+void
+IntegralReportingIDVPP::finalize()
+{
+  for (unsigned int ivar = 0; ivar < _nvar; ++ivar)
+    gatherSum(*_var_integrals[ivar]);
```
```suggestion
  for (auto & var_integral : _var_integrals)
    gatherSum(*var_integral);
```

loganharbour commented on modules/reactor/test/tests/meshgenerators/reporting_id/cartesian_id/tests:
```diff
@@ -0,0 +1,42 @@
+[Tests]
+  design = 'meshgenerators/CartesianIDPatternedMeshGenerator.md'
+  issues = '#19217'
+  [assembly.cell]
```
why is recover skipped here?

loganharbour commented on modules/reactor/test/tests/meshgenerators/reporting_id/depletion_id/tests:
```diff
@@ -0,0 +1,18 @@
+  design = 'meshgenerators/DepletionIDGenerator.md'
+  issues = '#19217'
+  [2dcore]
+    type = 'Exodiff'
```
same question about recover

loganharbour commented on modules/reactor/test/tests/meshgenerators/reporting_id/hexagonal_id/tests:
```diff
@@ -0,0 +1,42 @@
+[Tests]
+  design = 'meshgenerators/HexIDPatternedMeshGenerator.md'
+  issues = '#19217'
+  [assembly.cell]
```
same question about recover

yjung-anl replied:
@loganharbour 
What is the purpose of `recover` here?
This tests file was made based on the existing tests files in the reactor module, which set `recover` as `false`.

GiudGiud replied:
`recover` is a class of tests that run for half their timesteps, create a checkpoint, stop then recover from the checkpoint. 

Those tests were either steady state tests (which dont recover) or just mesh generation tests (also no recover)

yjung-anl commented:
Rebased this branch to accommodate the changes from PR #19358

yjung-anl committed:
removing duplicated functions of Cartesian and hexagonal reporting ID generators

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/891453/) on c3f8a6d : canceled by @yjung-anl

yjung-anl commented:
- Addressed all review comments. 
- Renamed the following two functions:
  - `IntegralReportingIDVPP` -> `ExtraIDIntegralVectorPostprocessor`
  - `PlaneIDGenerator` -> `PlaneIDMeshGenerator`

yjung-anl commented:
I will drop `DepletionIDGenerator` and `ExtraIDIntegralVectorPostprocessor` from this PR and make separate PR with following updates:
- `ExtraIDIntegralVectorPostprocessor`: adding an additional version based on "reporter system"
- `DepletionIDGenerator`: Make a generalized version named `ExtraIDCombiationGenerator` and make a wrapper for delepletion ID.

yjung-anl committed:
moving PlaneIDMeshGenerator to framework, refs idaholab#19217

yjung-anl commented:
@loganharbour @GiudGiud 
Can you re-review my PR?
I dropped `DepletionIDGenerator` and `ExtraIDIntegralVectorPostprocessor` from this PR and make separate PR with following updates:
* `ExtraIDIntegralVectorPostprocessor`: make an additional version based on "reporter system"
* `DepletionIDGenerator`: Make a generalized version named `ExtraIDCombiationGenerator` and make a wrapper for delepletion ID.
Additionally, I relocated `PlaneIDMeshGenerator` to the framework side.

loganharbour commented on framework/src/meshgenerators/PlaneIDMeshGenerator.C:
```diff
@@ -0,0 +1,108 @@
+  else
+    eid = mesh->get_elem_integer_index(_element_id_name);
+
+  Real tol = getParam<Real>("tolerance");
```
```suggestion
  const Real tol = getParam<Real>("tolerance");
```

loganharbour commented on framework/src/meshgenerators/PlaneIDMeshGenerator.C:
```diff
@@ -0,0 +1,108 @@
+  Real tol = getParam<Real>("tolerance");
+  for (auto & elem : mesh->active_element_ptr_range())
+  {
+    int layer_id = getPlaneID(elem->centroid());
```
```suggestion
   const int layer_id = getPlaneID(elem->centroid());
```

loganharbour commented on framework/src/meshgenerators/PlaneIDMeshGenerator.C:
```diff
@@ -0,0 +1,108 @@
+}
+
+int
+PlaneIDMeshGenerator::getPlaneID(const Point & p) const
```
Some comments are needed here - this isn't the most intuitive to read

loganharbour commented on framework/src/meshgenerators/PlaneIDMeshGenerator.C:
```diff
@@ -0,0 +1,108 @@
+int
+PlaneIDMeshGenerator::getPlaneID(const Point & p) const
+{
+  if (_planes.size() == 0)
```
Why would anyone ever use this with no planes provided? If planes aren't provided, I think it should error in the constructor

loganharbour commented on modules/reactor/include/meshgenerators/ReportingIDGeneratorBase.h:
```diff
@@ -0,0 +1,68 @@
+/**
+ * A base class that contains common members for reporting ID generators.
+ */
+class ReportingIDGeneratorBase
```
This should probably be put in utils; it isn't actually a mesh generator and should just be a set of static methods, i.e.:

```
namespace ReportingIDGeneratorUtils
{
getCellwiseIntegerIDs(const std::vector<std::unique_ptr<ReplicatedMesh>> & meshes,
                      const std::vector<std::vector<unsigned int>> & pattern,
                      const bool use_exclude_id,
                      const std::vector<bool> & exclude_ids);
}

```

etc

yjung-anl replied:
Thank you for suggestion. I will update it accordingly.

loganharbour commented on modules/reactor/src/meshgenerators/CartesianIDPatternedMeshGenerator.C:
```diff
@@ -0,0 +1,88 @@
+    integer_ids = getManualIntegerIDs(_meshes, _pattern, _id_pattern);
+
+  unsigned int extra_id_index = 0;
+  if (!mesh->has_elem_integer(_element_id_name))
```
Wouldn't you want to error here if the elem integer already exists?

yjung-anl replied:
Perhaps, there could be a user who want to overwrite to an existing ID. Thus, I did not error out here.
I will add a warning message for this.

loganharbour commented on modules/reactor/src/meshgenerators/CartesianIDPatternedMeshGenerator.C:
```diff
@@ -0,0 +1,88 @@
+  auto mesh = PatternedMeshGenerator::generate();
+  // assumes that the entire mesh has elements of each individual mesh sequentially ordered.
+  std::vector<dof_id_type> integer_ids;
+  if (_assign_type == "cell")
```
There's a lot of duplication between here and the hex ID generator

It looks like this can also go into the utility

loganharbour commented on modules/reactor/src/meshgenerators/HexIDPatternedMeshGenerator.C:
```diff
@@ -0,0 +1,149 @@
+  _exclude_ids.resize(_input_names.size());
+  if (_use_exclude_id)
+  {
+    std::vector<MeshGeneratorName> exclude_id_name =
```
More duplicated code

yjung-anl replied:
@loganharbour 
Duplications in Cartesian and hexagonal ones were removed by moving the those duplicated code sections into the utils. Thank you for suggestions.

loganharbour commented on modules/reactor/test/tests/meshgenerators/reporting_id/cartesian_id/tests:
```diff
@@ -0,0 +1,42 @@
+[Tests]
```
Use requirement/detail description. These should also be more human readable; they should describe something that is achieved and not a copy pasta of an enum (i.e., something like assign_type=cell is bad)

Should probably be something like this:

```
[assembly]
  requirement = 'The system shall support the generation of pin reporting IDs for cartesian assemblies'
  [cell]
    ..
    detail = 'for each component in the lattice in sequential order'
  []
  [pattern]
    ..
    detail = 'based on the ID of the input tiles'
  []
  [manual]
    ..
    detail = 'based on user-defined mapping'
  []
[]
```

loganharbour commented on modules/reactor/test/tests/meshgenerators/reporting_id/hexagonal_id/tests:
```diff
@@ -0,0 +1,42 @@
+[Tests]
```
Same thing that I suggested for requirement/detail in the cartesian generator

loganharbour commented on test/tests/meshgenerators/plane_id_mesh_generator/tests:
```diff
@@ -0,0 +1,27 @@
+    type = 'Exodiff'
+    input = 'plane_id_pin3d.i'
+    exodiff = 'plane_id_pin3d_out.e'
+    requirement = 'The system shall generate plane reporting IDs for 3d single pin'
```
Missing a complete sentence

loganharbour commented on test/tests/meshgenerators/plane_id_mesh_generator/tests:
```diff
@@ -0,0 +1,27 @@
+    requirement = 'The system shall generate plane reporting IDs for 3d single pin'
+    recover = false
+  []
+  [cartesian_grid_2d_x_dir]
+    type = 'Exodiff'
+    input = 'plane_id_grid2d.i'
+    exodiff = 'plane_id_grid2d_x_dir_out.e'
+    cli_args = "Outputs/file_base='plane_id_grid2d_x_dir_out'"
+    requirement = 'The system shall generate plane reporting IDs for Cartesian 2D grid by taking x-axis as plane direction'
+    recover = false
+  []
+  [cartesian_grid_2d_y_dir]
+    type = 'Exodiff'
+    input = 'plane_id_grid2d.i'
+    exodiff = 'plane_id_grid2d_y_dir_out.e'
+    cli_args = "Mesh/plane_id_gen/plane_axis='y' Outputs/file_base='plane_id_grid2d_y_dir_out'"
+    requirement = 'The system shall generate plane reporting IDs for Cartesian 2D grid by taking y-axis as plane direction'
+    recover = false
+  []
```
Use requirement/detail syntax

loganharbour commented on modules/reactor/src/utils/ReportingIDGeneratorUtils.C:
```diff
@@ -0,0 +1,201 @@
+{
+  std::vector<dof_id_type> integer_ids;
+  dof_id_type id = 0;
+  for (MooseIndex(pattern) i = 0; i < pattern.size(); ++i)
+  {
+    for (MooseIndex(pattern[i]) j = 0; j < pattern[i].size(); ++j)
+    {
+      ReplicatedMesh & cell_mesh = *meshes[pattern[i][j]];
+      unsigned int n_cell_elem = cell_mesh.n_elem();
+      bool exclude_id = false;
+      if (use_exclude_id)
+        if (exclude_ids[pattern[i][j]])
+          exclude_id = true;
+      if (!exclude_id)
+      {
+        for (unsigned int k = 0; k < n_cell_elem; ++k)
+          integer_ids.push_back(id);
+        ++id;
+      }
+      else
+        for (unsigned int k = 0; k < n_cell_elem; ++k)
+          integer_ids.push_back(DofObject::invalid_id);
+    }
+  }
```
```suggestion
  for (MooseIndex(pattern) i = 0; i < pattern.size(); ++i)
    for (MooseIndex(pattern[i]) j = 0; j < pattern[i].size(); ++j)
    {
      const auto value = (exclude_id || exclude_ids[pattern[i][j]]) ? DofObject::invalid_id : id++;
      std::insert(integer_ids.end(), meshes[pattern[i][j]]->n_elem(), value);
    }
```

Just a bit of a simplification :-)

loganharbour commented on modules/reactor/src/utils/ReportingIDGeneratorUtils.C:
```diff
@@ -0,0 +1,201 @@
+    const std::vector<std::vector<unsigned int>> & pattern)
+{
+  std::vector<dof_id_type> integer_ids;
+  for (MooseIndex(pattern) i = 0; i < pattern.size(); ++i)
+  {
+    for (MooseIndex(pattern[i]) j = 0; j < pattern[i].size(); ++j)
+    {
+      ReplicatedMesh & cell_mesh = *meshes[pattern[i][j]];
+      unsigned int n_cell_elem = cell_mesh.n_elem();
+      for (unsigned int k = 0; k < n_cell_elem; ++k)
+        integer_ids.push_back(pattern[i][j]);
+    }
+  }
```
```suggestion
  for (MooseIndex(pattern) i = 0; i < pattern.size(); ++i)
    for (MooseIndex(pattern[i]) j = 0; j < pattern[i].size(); ++j)
      std::insert(integer_ids.end(), meshes[pattern[i][j]]->n_elem(), pattern[i][j]);
```

loganharbour commented on modules/reactor/src/utils/ReportingIDGeneratorUtils.C:
```diff
@@ -0,0 +1,201 @@
+    const std::vector<std::vector<dof_id_type>> & id_pattern)
+{
+  std::vector<dof_id_type> integer_ids;
+  for (MooseIndex(pattern) i = 0; i < pattern.size(); ++i)
+  {
+    for (MooseIndex(pattern[i]) j = 0; j < pattern[i].size(); ++j)
+    {
+      unsigned int id = id_pattern[i][j];
+      ReplicatedMesh & cell_mesh = *meshes[pattern[i][j]];
+      unsigned int n_cell_elem = cell_mesh.n_elem();
+      for (unsigned int k = 0; k < n_cell_elem; ++k)
+        integer_ids.push_back(id);
+    }
+  }
```
```suggestion
  std::vector<dof_id_type> integer_ids;
  for (MooseIndex(pattern) i = 0; i < pattern.size(); ++i)
    for (MooseIndex(pattern[i]) j = 0; j < pattern[i].size(); ++j)
      std::insert(integer_ids.end(), meshes[pattern[i][j]]->n_elem(), id_pattern[i][j]);
```

loganharbour commented on modules/reactor/src/utils/ReportingIDGeneratorUtils.C:
```diff
@@ -0,0 +1,201 @@
+    const std::vector<std::vector<unsigned int>> & pattern)
+{
+  std::set<SubdomainID> blks;
+  for (MooseIndex(pattern) i = 0; i < pattern.size(); ++i)
+  {
+    for (MooseIndex(pattern[i]) j = 0; j < pattern[i].size(); ++j)
+    {
+      ReplicatedMesh & cell_mesh = *meshes[pattern[i][j]];
+      for (auto elem : cell_mesh.element_ptr_range())
+      {
+        auto blk = elem->subdomain_id();
+        auto it = blks.find(blk);
+        if (it == blks.end())
+          blks.insert(blk);
+      }
+    }
+  }
```
```suggestion
  for (MooseIndex(pattern) i = 0; i < pattern.size(); ++i)
    for (MooseIndex(pattern[i]) j = 0; j < pattern[i].size(); ++j)
    {
      std::set<SubdomainID> mesh_blks;
      meshes[pattern[i][j]]->subdomain_ids(mesh_blks);
      blks.insert(mesh_blks.begin(), mesh_blks.end());
    }
```

loganharbour commented on modules/reactor/src/utils/ReportingIDGeneratorUtils.C:
```diff
@@ -0,0 +1,201 @@
+                                           const std::set<SubdomainID> & blks)
+{
+  std::map<SubdomainID, unsigned int> blks_duct;
+  if (has_assembly_duct)
+  {
+    unsigned int i = 0;
+    for (auto elem : mesh->element_ptr_range())
+    {
+      auto blk = elem->subdomain_id();
+      auto it1 = blks.find(blk);
+      if (it1 == blks.end())
+      {
+        auto it2 = blks_duct.find(blk);
+        if (it2 == blks_duct.end())
+          blks_duct[blk] = i++;
+      }
+    }
+    blks_duct.erase(blks_duct.begin());
+  }
```
```suggestion
  if (has_assembly_duct)
  {
    std::set<SubdomainID> mesh_blks;
    mesh->subdomain_ids(mesh_blks);

    unsigned int i = 0;
    for (const auto mesh_blk : mesh_blks)
      if (!blks.count(mesh_blk))
        blks_duct[blk] = i++;

    // why do you do this? I don't know
    blks_duct.erase(blks_duct.begin());
  }
```

yjung-anl replied:
```
 blks_duct.erase(blks_duct.begin());
```
The above line is needed to delete the first entry because it is for surrounding regions between the assembly duct and pin (red block in the below figure).

![image](https://user-images.githubusercontent.com/73304529/144368900-89013a4f-b2c8-47d9-b7a0-d54b25d66db6.png)

Comments will be added

loganharbour commented on modules/reactor/src/utils/ReportingIDGeneratorUtils.C:
```diff
@@ -0,0 +1,201 @@
+  if (has_assembly_duct)
+  {
+    // setup assembly duct information
+    std::set<SubdomainID> blks = ReportingIDGeneratorUtils::getCellBlockIDs(input_meshes, pattern);
+    unsigned int duct_boundary_id = *std::max_element(integer_ids.begin(), integer_ids.end()) + 1;
+    std::map<SubdomainID, unsigned int> blks_duct =
```
```suggestion
    const std::set<SubdomainID> blks = ReportingIDGeneratorUtils::getCellBlockIDs(input_meshes, pattern);
   const unsigned int duct_boundary_id = *std::max_element(integer_ids.begin(), integer_ids.end()) + 1;
    const std::map<SubdomainID, unsigned int> blks_duct =
```

loganharbour commented on modules/reactor/src/utils/ReportingIDGeneratorUtils.C:
```diff
@@ -0,0 +1,201 @@
+    unsigned old_id = id;
+    for (auto elem : mesh->element_ptr_range())
+    {
+      auto blk = elem->subdomain_id();
+      auto it = blks.find(blk);
+      // check whether the current element belongs to duct/surrouding regions or not
+      if (it == blks.end())
+      {
+        // check whether the current element belongs to surroudning or duct regions
+        auto it2 = blks_duct.find(blk);
+        if (it2 == blks_duct.end())
+          // if the current element belongs to the surronding region
+          elem->set_extra_integer(extra_id_index, old_id);
+        else
+          // if the current element belongs to the duct region
+          elem->set_extra_integer(extra_id_index, duct_boundary_id + it2->second);
```
```suggestion
      const auto blk = elem->subdomain_id();
      // check whether the current element belongs to duct/surrouding regions or not
      if (!blks.count(blk))
      {
        // check whether the current element belongs to surroudning or duct regions
        if (!blks_duct.count(blk))
          // if the current element belongs to the surronding region
          elem->set_extra_integer(extra_id_index, old_id);
        else
          // if the current element belongs to the duct region
          elem->set_extra_integer(extra_id_index, duct_boundary_id + it2->second);
      }
```

loganharbour commented on modules/reactor/test/tests/meshgenerators/reporting_id/cartesian_id/tests:
```diff
@@ -0,0 +1,48 @@
+  design = 'meshgenerators/CartesianIDPatternedMeshGenerator.md'
+  issues = '#19217'
+  [assembly]
+    requirement = 'The system shall support the generation of reporting IDs for Cartesian assemblies'
```
```suggestion
    requirement = 'The system shall support the generation of reporting IDs for Cartesian assemblies assigned'
```

loganharbour commented on modules/reactor/test/tests/meshgenerators/reporting_id/cartesian_id/tests:
```diff
@@ -0,0 +1,48 @@
+      input = 'assembly_reporting_id.i'
+      exodiff = 'assembly_reporting_id_cell_out.e'
+      cli_args = "Outputs/file_base='assembly_reporting_id_cell_out'"
+      detail = 'assign pin IDs in sequential order for each components in the lattice'
```
```suggestion
      detail = 'in sequential order for each components in the lattice'
```

loganharbour commented on modules/reactor/test/tests/meshgenerators/reporting_id/cartesian_id/tests:
```diff
@@ -0,0 +1,48 @@
+      input = 'assembly_reporting_id.i'
+      exodiff = 'assembly_reporting_id_pattern_out.e'
+      cli_args = "Mesh/assembly/assign_type='pattern' Outputs/file_base='assembly_reporting_id_pattern_out'"
+      detail = 'assign pin IDs based on the ID of the input pattern tiles'
```
```suggestion
      detail = 'based on the ID of the input pattern tiles'
```

loganharbour commented on modules/reactor/test/tests/meshgenerators/reporting_id/cartesian_id/tests:
```diff
@@ -0,0 +1,48 @@
+      input = 'assembly_reporting_id.i'
+      exodiff = 'assembly_reporting_id_manual_out.e'
+      cli_args = "Mesh/assembly/assign_type='manual' Mesh/assembly/id_pattern='0 0 1 1;0 0 1 1;2 2 3 3;2 2 3 3' Outputs/file_base='assembly_reporting_id_manual_out'"
+      detail = 'assign pin IDs based on user-defined mapping'
```
```suggestion
      detail = 'based on user-defined mapping'
```

loganharbour commented on modules/reactor/test/tests/meshgenerators/reporting_id/cartesian_id/tests:
```diff
@@ -0,0 +1,48 @@
+    []
+  []
+  [core]
+    requirement = 'The system shall support the generation of reporting IDs for Cartesian cores'
```
```suggestion
    requirement = 'The system shall support the assignment of pin and assembly reporting IDs in a cartesian core with'
```

loganharbour commented on modules/reactor/test/tests/meshgenerators/reporting_id/cartesian_id/tests:
```diff
@@ -0,0 +1,48 @@
+      type = 'Exodiff'
+      input = 'core_reporting_id.i'
+      exodiff = 'core_reporting_id_out.e'
+      detail = 'assign pin and assembly reporting IDS for each pin and assembly in the Cartesian core'
```
Can you help me with this and the next test? I can't easily tell the difference between this test and the next one and we do need to distinguish them clearly

yjung-anl replied:
The first one has regular square core boundary.

The second one has zagged boundary that is formed by removal of certain assemblies on periphery.

loganharbour commented on modules/reactor/test/tests/meshgenerators/reporting_id/hexagonal_id/tests:
```diff
@@ -0,0 +1,48 @@
+  design = 'meshgenerators/HexIDPatternedMeshGenerator.md'
+  issues = '#19217'
+  [assembly]
+    requirement = 'The system shall support the generation of reporting IDs for hexagonal assemblies'
```
```suggestion
    requirement = 'The system shall support the generation of reporting IDs for hexagonal assemblies assigned'
```

loganharbour commented on modules/reactor/test/tests/meshgenerators/reporting_id/hexagonal_id/tests:
```diff
@@ -0,0 +1,48 @@
+      input = 'assembly_reporting_id.i'
+      exodiff = 'assembly_reporting_id_cell_out.e'
+      cli_args = "Outputs/file_base='assembly_reporting_id_cell_out'"
+      detail = 'assign pin IDs in sequential order for each components in the lattice'
```
```suggestion
      detail = 'in sequential order for each components in the lattice'
```

loganharbour commented on modules/reactor/test/tests/meshgenerators/reporting_id/hexagonal_id/tests:
```diff
@@ -0,0 +1,48 @@
+      input = 'assembly_reporting_id.i'
+      exodiff = 'assembly_reporting_id_pattern_out.e'
+      cli_args = "Mesh/assembly/assign_type='pattern' Outputs/file_base='assembly_reporting_id_pattern_out'"
+      detail = 'assign pin IDs based on the ID of the input pattern tiles'
```
```suggestion
      detail = 'based on the ID of the input pattern tiles'
```

loganharbour commented on modules/reactor/test/tests/meshgenerators/reporting_id/hexagonal_id/tests:
```diff
@@ -0,0 +1,48 @@
+      input = 'assembly_reporting_id.i'
+      exodiff = 'assembly_reporting_id_manual_out.e'
+      cli_args = "Mesh/assembly/assign_type='manual' Mesh/assembly/id_pattern='2 2 2;2 1 1 2;2 1 0 1 2;2 1 1 2;2 2 2' Outputs/file_base='assembly_reporting_id_manual_out'"
+      detail = 'assign pin IDs based on user-defined mapping'
```
```suggestion
      detail = 'based on user-defined mapping'
```

loganharbour commented on modules/reactor/test/tests/meshgenerators/reporting_id/hexagonal_id/tests:
```diff
@@ -0,0 +1,48 @@
+      recover = false
+    []
+  []
+  [core]
```
I think these can probably be split up with just two requirements and detail, but I don't completely follow what they're testing

loganharbour reviewed (changes requested):
Okay, last around! Few simple cleanups that helped me understand what was going on more. The generalization into the utility looks great. Once these are satisfied, I'm good for a 👍🏼

yjung-anl commented on modules/reactor/src/meshgenerators/HexIDPatternedMeshGenerator.C:
```diff
@@ -0,0 +1,103 @@
+    mooseWarning("extra integer id name already exists", _element_id_name);
+  }
+
+  std::vector<std::unique_ptr<ReplicatedMesh>> meshes;
+  meshes.reserve(_input_names.size());
+  for (MooseIndex(_input_names) i = 0; i < _input_names.size(); ++i)
+  {
+    std::unique_ptr<ReplicatedMesh> cell_mesh =
+        dynamic_pointer_cast<ReplicatedMesh>(*_mesh_ptrs[i]);
+    meshes.push_back(std::move(cell_mesh));
+  }
```
I tried to pass `const std::vector<std::unique_ptr<MeshBase> *> _mesh_ptrs` directly to the util function.
But the code crashed for this operation: `_mesh_ptrs[pattern[i][j]]->n_elem()`.
Perhaps, do you have any suggestion to access individual mesh data in `_mesh_ptrs`?

loganharbour replied:
Oh... duh. I missed that one was a pointer to unique_ptrs and the other one was just unique_ptrs. Ignore this; we're good!

yjung-anl commented:
@loganharbour 
I addressed all of your review comments. Thank you again for your review!!

loganharbour reviewed (approved):
@yjung-anl thanks for all of your patience working through this!

eshemon commented:
Thank you very much, @loganharbour!
